File: js/src/orchestration/result.ts
import type { NaruResult } from "../types.js";
import type { AgentDecisionTrace } from "./trace.js";
import type { OrchestratorIntent } from "./intent.js";

export interface PendingConfirmation {
  type: string;
  payload: Record<string, unknown>;
}

/**
 * OrchestrationResult extends NaruResult with orchestration-specific fields.
 * All new fields are optional or have reasonable defaults.
 *
 * Note: `trace` from NaruResult (Trace | null) is preserved for backward compat.
 * The orchestration decision trace is exposed via `decisionTrace`.
 * The orchestration-resolved intent is exposed via `orchestrationIntent`.
 */
export interface OrchestrationResult extends NaruResult {
  /** Full decision trace for this orchestration invocation */
  decisionTrace: AgentDecisionTrace;
  /** If set, agent is waiting for user confirmation */
  pendingConfirmation: PendingConfirmation | null;
  /** The orchestration-resolved intent (null if no resolver configured) */
  orchestrationIntent: OrchestratorIntent<string> | null;
}

File: naru_agent/orchestration/result.py
"""OrchestrationResult — extends NaruResult with orchestration-specific fields."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any

from naru_agent.agent import NaruResult

if TYPE_CHECKING:
    from naru_agent.orchestration.intent import OrchestratorIntent
    from naru_agent.orchestration.trace import AgentDecisionTrace


@dataclass
class PendingConfirmation:
    """Represents a pending confirmation awaiting user response."""

    type: str  # noqa: A003
    payload: dict[str, Any] = field(default_factory=dict)


@dataclass
class OrchestrationResult(NaruResult):
    """Extends NaruResult with orchestration-specific fields.

    All new fields are optional or have reasonable defaults so existing
    NaruResult consumers continue to work (backward compatibility).
    """

    decision_trace: AgentDecisionTrace | None = None  # type: ignore[assignment]
    pending_confirmation: PendingConfirmation | None = None
    orchestration_intent: OrchestratorIntent | None = None  # type: ignore[type-arg]
